### PR TITLE
[thermo] `RedlichKwongMFTP` critical property lookups

### PR DESCRIPTION
**Pull request by decaluwe**
Changes proposed in this pull request:
-Makes it so that `Thermo::RedlichKwongMFTP` input files do not necessarily have to provide `pureFluidParameters` for every species.  Since these can be calculated from the critical temperature and pressure for the Redlich Kwong EoS, we provide the ability to look up Tc and Pc for well-known species.
-Provides a database of species with well-known critical properties.  Species are currently matched by chemical formula, so isomers are currently prohibited.  Can possibly add functionality to scan by species name, but this seems like it could lead to significant ambiguity.  My preference is to keep this database limited to species where potential ambiguity is near zero.  Database is stored in the `data` folder.  Updating the database is relatively straight-foward for any end users - user simply has to lookup Tc and Pc and enter them into the database, rather than lookup Tc and Pc, calculate a and b parameters for R-K EoS, and then update the cti file.
-Adds test suite coverage for this capability.

### CODE REVIEW

File: data/inputs/critProperties.xml [hunk 1]
```diff
@@ -0,0 +1,1150 @@
+<ctml>
```
Comment (speth) [inline data/inputs/critProperties.xml]: Is there a reference for these values?

File: data/inputs/critProperties.xml [hunk 2]
```diff
@@ -0,0 +1,1150 @@
...
+
+	   <species name=      "CCL4">
```
Comment (speth) [inline data/inputs/critProperties.xml]: Spaces, not tabs, please.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 3]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
     MixtureFugacityTP::initThermoXML(phaseNode, id);
 }
 
+doublereal* RedlichKwongMFTP::getCoeff(const std::string& iName)
+{
+    vector_fp vParams;
+    static doublereal spCoeff[2];
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: I would recommend against using `static` arrays, as they introduce potential race conditions. Better to have this function return a `vector<double>`.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 4]
```diff
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+                            a_coeff_vec(0, counter) = coeffArray[0];
+                            a_coeff_vec(1, counter) = 0.0;
+                            b_vec_Curr_[iSpecies] = coeffArray[1];
+                            
+                            setSpeciesCoeffs(iName, a_coeff_vec(0, counter),
+                                             a_coeff_vec(1, counter),
+                                             b_vec_Curr_[iSpecies]);
+                        }
+                        else
+                        {
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: Prefer writing this on one line, e.g.
```
    end block 1;
} else {
    start block 2;
```

File: src/thermo/RedlichKwongMFTP.cpp [hunk 5]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+    XML_Node* doc = get_XML_File("critProperties.xml");
+    size_t nDatabase = doc->nChildren();
+    XML_Node* acNodeDocPtr = 0;
+
+    // Loop through all species in the database and attempt to match supplied species to each.
+    // If present, calculate pureFluidParameters a_k and b_k based on crit properties T_c and P_c:
+    for (size_t isp = 0; isp < nDatabase; isp++)
+    {
+        XML_Node& acNodeDoc = doc->child(isp);
+        acNodeDocPtr = &acNodeDoc;
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: Doesn't look like this variable ever gets used.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 6]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+        std::string dbName = ba::to_lower_copy(acNodeDoc.attrib("name"));
+
+        // Attempt to match provided specie iName to current database specie dbName:
+        if (iName == dbName)
+        {
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: Prefer to write this on one line, e.g.
```c++
if (condition) {
    block;
```

File: src/thermo/RedlichKwongMFTP.cpp [hunk 7]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+            // Read from database and calculate a and b coefficients
+            doublereal vParams;
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: Prefer to just use `double` instead of the unnecessary typedef `doublereal`.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 8]
```diff
@@ -77,9 +77,19 @@ void RedlichKwongMFTP::setSpeciesCoeffs(const std::string& species,
...
+        // a_coeff_vec is initialized to -1, so screen for unidentified species to prevent
+        // imaginary numbers on the off-diagonals:
+        if (a_coeff_vec(0, j + m_kk * j) < 0 || a_coeff_vec(1, j + m_kk * j) < 0)
+        {
+            // The diagonal element of the jth species has not yet been defined.
+            continue;
+        }
+        // Only use the mixing rules if the off-diagonal element has not already been defined by a
+        // user-specified crossFluidParameters entry:
+        else if (a_coeff_vec(0, j + m_kk * k) == -1) {
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: braces and `else if` on one line, with comment inside the latter code block

File: src/thermo/RedlichKwongMFTP.cpp [hunk 9]
```diff
@@ -544,9 +554,11 @@ bool RedlichKwongMFTP::addSpecies(shared_ptr<Species> spec)
     bool added = MixtureFugacityTP::addSpecies(spec);
     if (added) {
         a_vec_Curr_.resize(m_kk * m_kk, 0.0);
-        b_vec_Curr_.push_back(0.0);
 
-        a_coeff_vec.resize(2, m_kk * m_kk, 0.0);
+        // Initialize a_vec and b_vec to -1, to screen for species with pureFluidParameters which
+        //     undefined in the input file:
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: "are undefined"?

File: src/thermo/RedlichKwongMFTP.cpp [hunk 10]
```diff
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+            size_t nParams = 0;
+
+            // Loop through the children and read out fluid parameters.  Process all the
+            // pureFluidParameters, first:
             for (size_t i = 0; i < acNode.nChildren(); i++) {
                 XML_Node& xmlACChild = acNode.child(i);
                 if (ba::iequals(xmlACChild.name(), "purefluidparameters")) {
                     readXMLPureFluid(xmlACChild);
-                } else if (ba::iequals(xmlACChild.name(), "crossfluidparameters")) {
+                    nParams += 1;
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: The existence of N `purefluidparameters` nodes for a phase with N species isn't necessarily a guarantee that there are parameters for each species -- there could be duplicates or parameters for species that aren't in the phase. It's safer and easy enough to always loop over all the species and make sure they don't have their `a` values set to -1.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 11]
```diff
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+                getStringArray(sparrays, spnames);
+
+                // Loop through all species in the CTI file
+                size_t iSpecies = 0;
+
+                for (size_t i = 0; i < m_kk; i++)
+                {
+                    string iName = ba::to_lower_copy(spnames[i]);
+
+                    // Get the index of the species
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: Rather than getting the species names from the `speciesArray` XML node, since they've all already been added to the phase, you can just loop over `k = 0; k < m_kk` and call `speciesName(k)` to get the name.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 12]
```diff
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+                    // If not, then search the database:
+                    if (a_coeff_vec(0, counter) == -1 ||
+                        b_vec_Curr_[iSpecies] == -1) {
+
+                        doublereal *coeffArray;
+
+                        // Search the database for the species name and calculate coefficients
+                        // a and b, from critical properties:
+                        // coeffArray[0] = a0, coeffArray[1] = b;
+                        coeffArray = getCoeff(iName);
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: If not finding coefficients for a species either in the phase definition or the database is considered an error, would it make more sense to just have `getCoeff` raise an exception if the species is missing, rather than returning a special value and then checking for that value here? I think this would also simplify the implementation of `getCoeff`.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 13]
```diff
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+
+                        // Check if species was found in the database of critical properties,
+                        // and either assign the results or throw an error:
+                        if (coeffArray[0] != -1 || coeffArray[1] != -1) {
+                            //Assuming no temperature dependence (i,e a1 = 0)
+                            a_coeff_vec(0, counter) = coeffArray[0];
+                            a_coeff_vec(1, counter) = 0.0;
+                            b_vec_Curr_[iSpecies] = coeffArray[1];
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: The entries in `a_coeff_vec` and `b_vec_Curr_` don't need to be set here -- that's done by `setSpeciesCoeffs`.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 14]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+    // Get number of species in the databae
+    // open xml file critProperties.xml
+    XML_Node* doc = get_XML_File("critProperties.xml");
+    size_t nDatabase = doc->nChildren();
+    XML_Node* acNodeDocPtr = 0;
+
+    // Loop through all species in the database and attempt to match supplied species to each.
+    // If present, calculate pureFluidParameters a_k and b_k based on crit properties T_c and P_c:
+    for (size_t isp = 0; isp < nDatabase; isp++)
+    {
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: Bracket on previous line, please.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 15]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
 }
 
+doublereal* RedlichKwongMFTP::getCoeff(const std::string& iName)
+{
+    vector_fp vParams;
+    static doublereal spCoeff[2];
+    spCoeff[0] = -1;
+    spCoeff[1] = -1;
+
+    // Get number of species in the databae
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: 'database' is misspelled

File: src/thermo/RedlichKwongMFTP.cpp [hunk 16]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+                {
+                    throw CanteraError("RedlichKwongMFTP::GetCoeff",
+                                       "Critical Pressure must be positive ");
+                }
+                P_crit = vParams;
+            }
+
+            //Assuming no temperature dependence
+            //Pc from tha database is in bars. convert to Pascals
+            P_crit = P_crit * 100000;
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: I would rather the database didn't have values in non-SI units, especially if they aren't going to be labeled values.

File: src/thermo/RedlichKwongMFTP.cpp [hunk 17]
```diff
@@ -587,6 +666,78 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
...
+            
+            spCoeff[0] = 0.42748 * pow(GasConstant, 2) * pow(T_crit, 2.5) / P_crit; //coeff a
+            spCoeff[1] = 0.08664 *GasConstant * T_crit / P_crit; // coeff b
```
Comment (speth) [inline src/thermo/RedlichKwongMFTP.cpp]: minor nitpick: space on both sides of the `*`.

File: src/thermo/critProperties.xml [hunk 18]
```diff
@@ -0,0 +1,3 @@
+<ctml>
```
Comment (speth) [inline src/thermo/critProperties.xml]: I think this file should be deleted (preferably, do this as a separate commit, then do an interactive rebase and squash that commit with the one that introduced the file in this location, so that it's like it never existed)

File: test/thermo/RedlichKwongMFTP_Test.cpp [hunk 19]
```diff
@@ -26,12 +27,16 @@ class RedlichKwongMFTP_Test : public testing::Test
 
 TEST_F(RedlichKwongMFTP_Test, construct_from_cti)
 {
+    const std::string valid_file("../data/co2_RK_example.cti");
+    initializeTestPhase(valid_file);
```
Comment (speth) [inline test/thermo/RedlichKwongMFTP_Test.cpp]: I think it's fine to just write `initializeTestPhase("../data/co2_RK_example.cti");`

File: test/thermo/RedlichKwongMFTP_Test.cpp [hunk 20]
```diff
@@ -9,8 +9,9 @@ namespace Cantera
 class RedlichKwongMFTP_Test : public testing::Test
 {
 public:
-    RedlichKwongMFTP_Test() {
-        test_phase.reset(newPhase("../data/co2_RK_example.cti"));
+    void initializeTestPhase(const std::string & filename)
```
Comment (speth) [inline test/thermo/RedlichKwongMFTP_Test.cpp]: Our normal style is to not have the space between the typename and the `&`, e.g. `std::string& filename`.

File: test/thermo/RedlichKwongMFTP_Test.cpp [hunk 21]
```diff
@@ -165,4 +178,29 @@ TEST_F(RedlichKwongMFTP_Test, setTP)
         EXPECT_NEAR(test_phase->density(),p3[i],1.e-8);
     }
 }
+
+TEST_F(RedlichKwongMFTP_Test, critPropLookup)
+{
+    // Check to make sure that RedlichKwongMFTP is able to properly calcualte a and b
```
Comment (speth) [inline test/thermo/RedlichKwongMFTP_Test.cpp]: "calculate"

**Review by speth — CHANGES_REQUESTED** [review]
This looks really interesting, and will certainly make the RedlichKwong class much easier to use!

Based on the implementation, I would say that species are matched by name (identifier), not formula (composition), which does allow you to define isomers, but with no standard naming scheme. In order to expand the database and eliminate ambiguities, I think we would need to have a better identifier present, e.g. something like InChI. Not something that we need to worry about now, but worth considering for the future.

Rather than having users update this database, the other option for providing Tc/Pc instead of the `a` and `b` coefficients would be to look for this information in the species declaration as well. I think this would be worth looking at after we get a bit further with the new input file format.

File: src/thermo/RedlichKwongMFTP.cpp (hunk 11 above)
Comment (decaluwe) [reply src/thermo/RedlichKwongMFTP.cpp]: Hey, finally working on implementing these, and for some reason this one change does not work.  The lookup function fails when I make this change.

Separate (but possibly related) question: do I actually need to do this circular type loop to then retrieve the species index?  if I'm looping over `k = 0; k < m_kk`, is k going to be equal to the iSpecies that I calculate on line 616?

Comment (speth) [reply src/thermo/RedlichKwongMFTP.cpp]: Right, you don't need to call `speciesIndex` to get the index, since you are already just looping over the species (so the `iSpecies` variable should be eliminated). What I am suggesting is that you also shouldn't need to touch the `phaseNode` to get at the species names -- `spnames[i]` should be the same as just calling `speciesName(i)`.

**Comment by speth** [comment]
Before making any more changes on this branch, I think we should address the warning regarding conflicts and rebase this PR on top of the current master. I have already done this if you want to pull the branch https://github.com/speth/cantera/tree/critProps and reset this PR to that commit.

**Comment by decaluwe** [comment]
Hi Ray, 

Yes, I’ve already done this on my local branch. 

On the speciesName(i) call, what I’m saying is that making that change does not work. The critPropLookup test fails (this is replicated after I install, just trying to load a cti with some missing pureFluidParemeters - the critical properties are not successfully read from the xml.) 

Have not been able to dig in and figure out why, but it seems odd that it would not work. 

Steven

**Comment by speth** [comment]
That's strange. What do you get if you use `writelog` to print out the species name from each source as you go through the loop? If you want to push a branch showing the problem, I can take a look at it.
